Ticket 35059 - call response.close, or failing that request_finished.send, if processing is cancelled

jthorniley opened the pull request:
The ASGIHandler class attempts to cancel view processing when the http connection is closed. This could happen at any point and potentially also cancels the `request.close` cleanup code.

This moves the call to `request.close` outside of the cancellable `send_response` function, and additionally wraps it in `asyncio.shield`.

For consistency, this also invokes the `request_finished` signal if async view processing is cancelled by this method before a response object has been created. This is because the view could have created DB connections during processing and it if the `request_finished` signal is not sent they may not be cleaned up.

nessita reviewed (commented):
I branched the PR and reviewed the code, focusing on the tests. The moved/copied code for closing the response added in the `if request is None` branch is not properly tested (in that if I remove that added line, no test fail).

Could you try to add a test or a set of assertions that ensure this needed code is not removed?

carltongibson commented on django/core/handlers/asgi.py:
```diff
@@ -223,6 +227,17 @@ async def process_request(request, send):
+            # object is created. We still have to fire the request_finished signal to
+            # close resources that might have been opened while the request was
+            # processing
+            signals.request_finished.send(sender=self)
```
This should probably use `asend` no? 

Q: Would this be better placed in the `# Ignore client disconnects.` exception handler (where we currently `pass`)? 🤔

jthorniley replied:
`asend` - yes - I've made that update and also modified the tests so they are more strict about this

carltongibson commented on django/core/handlers/asgi.py:
```diff
@@ -186,9 +186,13 @@ async def handle(self, scope, receive, send):
         if request is None:
             body_file.close()
             await self.send_response(error_response, send)
+            await sync_to_async(error_response.close, thread_sensitive=True)()
```
Is this safe to remove? 🤔 (We `return` here, so when is `error_response.close` called?)

jthorniley replied:
Interesting question - I'm not sure, my only reason to add it was that this is code was in send_response, and therefore it would have been called before this change. While I don't know of any reason why the django code needs it, the `request_finished` signal is documented so applications can use it, and I guess could in theory depend on this behaviour?

carltongibson replied:
If it's me, it's a separate ticket. 

* Is it an issue?
* If so, can we demonstrate that in a test. 

But either way, it's not the issue here is it?

jthorniley replied:
Ok, I think the conclusion we are coming to is this is not needed - there's no test cases that currently fail and its not obvious that it is needed. I can remove.

felixxm replied:
This PR is intended for backporting, so we don't want to make any unrelated changes in it. Please revert, and feel-free to send a follow up PR when this one is merged.

carltongibson replied:
Reviewing this conversation now, I'm confused by my initial comment. 

I think this **is** needed 

@jthorniley's: 

> ...reason to add it was that this is code was in send_response, and therefore it would have been called before this change.

That sounds right. 

Sorry if I confused the discussion there. (I can't recall exactly what I thought at the time.)

carltongibson commented on django/core/handlers/asgi.py:
```diff
@@ -223,6 +227,17 @@ async def process_request(request, send):
+        if response is not None:
+            # If response exists, we must ensure `close` gets called. Do not allow this
+            # to be cancelled
+            await asyncio.shield(sync_to_async(response.close, thread_sensitive=True)())
```
If we handle the `RequestAborted` case, do we need to move this here? 

i.e. I see the failure to dispatch `request_finished` on the cancelled case, but not (yet in tests at least) otherwise 🤔

carltongibson reviewed (commented):
Hey @jthorniley — thanks for the investigation. This seems plausible to me, yes. 

Can I let you resolve @nessita's queries, and then I can take another look if desired?

A couple of points: 

* As per the inline comments, can we reduce it to just making sure we dispatch the signal in the disconnect case? (That would avoid the nonlocal usage and the keep the flows for each branch more closely related). 
* Comments should avoid use of "we". 
* Would need a release note in 5.0.x.txt.

carltongibson commented:
Just to rephrase, slightly... 

> … can we reduce it to just making sure we dispatch the signal in the disconnect case? 

The tests you added pass with this diff (from `main`): 

```diff
diff --git a/django/core/handlers/asgi.py b/django/core/handlers/asgi.py
index 7b0086fb76..eef8bf4e7c 100644
--- a/django/core/handlers/asgi.py
+++ b/django/core/handlers/asgi.py
@@ -209,7 +209,7 @@ class ASGIHandler(base.BaseHandler):
                     task.result()
                 except RequestAborted:
                     # Ignore client disconnects.
-                    pass
+                    signals.request_finished.send(sender=self)
                 except AssertionError:
                     body_file.close()
                     raise
```

jthorniley commented:
Thanks both @nessita @carltongibson for your comments I will take a look today and respond but this all sounds reasonable to me.

jthorniley commented:
> I branched the PR and reviewed the code, focusing on the tests. The moved/copied code for closing the response added in the `if request is None` branch is not properly tested (in that if I remove that added line, no test fail).
> 
> Could you try to add a test or a set of assertions that ensure this needed code is not removed?

@nessita I think this fails the `asgi.tests.ASGITest.test_asyncio_cancel_error` - if I'm reading this right you mean the `else` branch - so in the current version if I make this change then I get a test failure:

```
diff --git a/django/core/handlers/asgi.py b/django/core/handlers/asgi.py
index 295e546053..931e81df61 100644
--- a/django/core/handlers/asgi.py
+++ b/django/core/handlers/asgi.py
@@ -236,7 +236,7 @@ class ASGIHandler(base.BaseHandler):
             # run_get_response, so we cannot call response.close
             # However, processing could have acquired resources that need
             # to be released by handling the request_finished signal
-            await signals.request_finished.asend(sender=self)
+            pass # await signals.request_finished.asend(sender=self)
 
     async def listen_for_disconnect(self, receive):
         """Listen for disconnect from the client."""
```

jthorniley commented:
> Just to rephrase, slightly...
> 
> > … can we reduce it to just making sure we dispatch the signal in the disconnect case?
> 
> The tests you added pass with this diff (from `main`):
> 
> ```diff
> diff --git a/django/core/handlers/asgi.py b/django/core/handlers/asgi.py
> index 7b0086fb76..eef8bf4e7c 100644
> --- a/django/core/handlers/asgi.py
> +++ b/django/core/handlers/asgi.py
> @@ -209,7 +209,7 @@ class ASGIHandler(base.BaseHandler):
>                      task.result()
>                  except RequestAborted:
>                      # Ignore client disconnects.
> -                    pass
> +                    signals.request_finished.send(sender=self)
>                  except AssertionError:
>                      body_file.close()
>                      raise
> ```
> 
> Update: as per my own comment 🤦‍♀️, this also passes, and is probably what we want here: `await signals.request_finished.asend(sender=self)` 🤔

I'm going on my best understanding here so happy to be corrected, from what I see `http.disconnect` (looking at the ASGI spec) should be returned from `await receive()` when either:

* the client disconnects early, OR
* when the http request/response cycle is complete and the ASGI server is not expecting any more `send` calls for this request

So, in that second case, generally for non-streaming responses, the `process_request` task can complete before `http.disconnect` arrives - in that case the `listen_for_disconnect` task can be the one that gets cancelled before `RequestAborted` is raised.

I've added another test case, `test_request_lifecycle_signals_do_not_require_http_disconnect` - that simulates this path. I'm not sure how artificial it is, but I feel like there's definitely a potential race condition here.

I've modified the existing test case `test_request_lifecycle_signals_dispatched_with_thread_sensitive` so that the `http.disconnect` is received first. 

I think you should find with these test cases that the proposed patch no longer works, because it will cause the `request_finished` signal to be triggered twice - unless you remove the call to response.close from `send_response`,  but in that case `request_finished` may never get called in that second scenario I describe.

I've updated the new tests to require `request_finished` to get invoked just once also. While you could argue that for cleanup code its only a minor performance issue if it gets called twice, you could imagine other uses (e.g. metrics) where it would be a undesirable change in behvaiour - I think in WSGI its pretty reliably called exactly once.

carltongibson commented on django/core/handlers/asgi.py:
```diff
@@ -223,6 +226,17 @@ async def process_request(request, send):
+            # signal and ensures expected cleanup code is called
+            await sync_to_async(response.close, thread_sensitive=True)()
+        else:
+            # response is None - process_request was cancelled during or before
+            # run_get_response, so we cannot call response.close
+            # However, processing could have acquired resources that need
+            # to be released by handling the request_finished signal
+            await signals.request_finished.asend(sender=self)
```
I think the comments can be reduced here. 

```suggestion
            # Client disconnected. 
            await signals.request_finished.asend(sender=self)
```

felixxm replied:
I think we should a class not an instance
```suggestion
            await signals.request_finished.asend(sender=self.__class__)
```

carltongibson reviewed (commented):
Hi @jthorniley — thanks for the follow-up. 

Let me look at the extra test cases. (It's Friday, and I need a ponder 💭) I agree the `request_finished` signal should only be dispatched once. 

Your _second case_ is the _happy path_: `process_request` finishes and we send the response. But I'm not expecting that we see `http.disconnect` at all in that case. (I will review though; not sure that it changes too much.)

carltongibson commented:
@jthorniley OK, great. That makes sense. Let me dig in and have a look at it. Good work! 🎁

carltongibson commented on tests/asgi/tests.py:
```diff
@@ -333,10 +335,51 @@ def __call__(self, **kwargs):
         response_body = await communicator.receive_output()
         self.assertEqual(response_body["type"], "http.response.body")
         self.assertEqual(response_body["body"], b"Hello World!")
+        # The application gets http.disconnect as the response is complete
+        await communicator.send_input({"type": "http.disconnect"})
```
This is the key bit. If the application server sends an `http.disonnect` before the `process_request` task completes and the `listen_for_disconnect` taks is processed first, then either we risk getting the (issue here) `request.close()` not called as `process_request` is cancelled, or it can be called twice (on different pathways) if we add that into the `try... except...` when cleaning up the tasks. 

I didn't manage to reproduce that, but it seems conceivable, yes.

jthorniley replied:
Yes its tricky to test. I think the problem is most acute where the cancellation arrives during the async code that is invoked by `await sync_to_async(request.close)()` - there isn't a easy hook to simulate that (the only async code that could be cancelled is the `sync_to_async` wrapper, not `request.close` itself as that is sync).

Having this line makes the test fail with the `await request_finished.asend()` in the `except RequestAborted` block approach that I tried last week. However, that also causes a test failure in test_asyncio_streamming_cancel_error now, so this can be safely removed I think.

The test below just sends http.disconnect in a different order, so again it fails under other patched but isn't currently being exercised, I think I can remove it.

carltongibson reviewed (commented):
Hey @jthorniley — thanks again for this. 

tl;dr — yes I think it makes sense. So, if you can adjust for the comments, I think it's probably correct. Calling `request.close` or dispatching `request_finished` after **both** tasks are completed would eliminate the potential race.

I wonder if we can tighten up the tests a bit. 🤔

@nessita / @felixxm — I'd need another quick look to be 100% on the `async_unsafe` and `nonlocal` usages in the tests. (I'm sure they're fine but is there a more straightforward way?) Happy to have another look, but, for me, this looks progress-able (modulo comments). 

Thanks again 🎁

carltongibson commented on tests/asgi/tests.py:
```diff
@@ -410,6 +461,8 @@ class TestASGIHandler(ASGIHandler):
         # Give response.close() time to finish.
         await communicator.wait()
         self.assertIs(view_did_cancel, False)
+        self.assertNotEqual(finished_threads.pop(), threading.current_thread())
+        self.assertEqual(finished_threads, [])
```
I'd probably assert `len(finished_threads)` here, maybe with a comment — i.e. that it was called once only (which is the intention right?)

carltongibson commented on tests/asgi/tests.py:
```diff
@@ -315,6 +316,7 @@ class SignalHandler:
 
             threads = []
 
+            @async_unsafe
```
Commenting this doesn't cause a failure.

nessita replied:
Changing the call `asend` to `send` does produce failures independently of whether this decorator is present or not. Are you certain is needed?

jthorniley replied:
Thanks - I have gone over this and confirmed its probably not needed. Since its the only change here I'm happy to remove. There is a failure if request.close() is called without the sync_to_async wrapper, but its not necessary to have this decorator as the thread tests also fail in that case.

jthorniley commented on tests/asgi/tests.py:
```diff
@@ -464,6 +529,8 @@ class TestASGIHandler(ASGIHandler):
         self.assertEqual(response_body["body"], b"Hello World!")
         await communicator.wait()
         self.assertIs(view_did_cancel, False)
+        self.assertNotEqual(finished_threads.pop(), threading.current_thread())
+        self.assertEqual(finished_threads, [])
```
I think in both cases I just wanted to be clear that we have exactly one call (so something along the lines of assertEqual(call_count, 1) is perhaps clearer.

felixxm commented on django/core/handlers/asgi.py:
```diff
@@ -223,6 +226,17 @@ async def process_request(request, send):
+        if response is not None:
+            # call response.close in sync thread - this invokes the request_finished
+            # signal and ensures expected cleanup code is called
+            await sync_to_async(response.close, thread_sensitive=True)()
```
`thread_sensitive=True` is the default you can remove it.

felixxm commented on docs/releases/5.0.2.txt:
```diff
@@ -9,4 +9,6 @@ Django 5.0.2 fixes several bugs in 5.0.1.
 ========
 
-* ...
+* Fixed a race condition which caused the ``request_finished`` signal to
```
```suggestion
* Fixed a regression in Django 5.0 that caused the ..
```

jthorniley commented:
I think I've addressed the comments. (Left commits un-squashed for review)

nessita commented:
> I think I've addressed the comments. (Left commits un-squashed for review)

Thank you! For future reference, when a PR is ready for a re-review, its status should be updated in the ticket following [these docs](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#patch-style) (basically, "patch needs improvement" should be unset in the ticket).

felixxm commented on django/core/handlers/asgi.py:
```diff
@@ -221,8 +233,15 @@ async def process_request(request, send):
                 except asyncio.CancelledError:
                     # Task re-raised the CancelledError as expected.
                     pass
+
```
Revert unrelated blank-line changes.

felixxm commented on tests/asgi/tests.py:
```diff
@@ -333,10 +335,11 @@ def __call__(self, **kwargs):
         response_body = await communicator.receive_output()
         self.assertEqual(response_body["type"], "http.response.body")
         self.assertEqual(response_body["body"], b"Hello World!")
+
```
Revert unrelated blank-line changes.

felixxm commented on docs/releases/5.0.2.txt:
```diff
@@ -12,3 +12,6 @@ Bugfixes
 * Reallowed, following a regression in Django 5.0.1, filtering against local
   foreign keys not included in :attr:`.ModelAdmin.list_filter`
   (:ticket:`35087`).
+* Fixed a regression in Django 5.0 that caused the ``request_finished`` signal to
```
Blank line is missing
```suggestion

* Fixed a regression in Django 5.0 that caused the ``request_finished`` signal to
```

felixxm commented on tests/asgi/tests.py:
```diff
@@ -425,11 +440,25 @@ class TestASGIHandler(ASGIHandler):
             await communicator.receive_output()
         await communicator.wait()
         self.assertIs(view_did_cancel, True)
+        # Assert exactly one call to on_finished handler
+        self.assertEqual(len(finished_threads), 1)
+        # Assert it was NOT on the async thread
```
To add more explanation, docstrings/comments should state the expected behavior and omit prefixes like "Assert" since all assertions assert things. This guideline is from [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).

felixxm commented on tests/asgi/tests.py:
```diff
@@ -333,10 +335,11 @@ def __call__(self, **kwargs):
         await communicator.wait()
 
-        # AsyncToSync should have executed the signals in the same thread.
+        # SyncToAsync should have executed the signals in the same thread.
```
If there is a typo here, it's unrelated with this patch. Please send a fix separately.

jthorniley replied:
ok yeah, I'm pretty sure there is (and I had previously modified this test, but I'll revert all the changes)

jthorniley replied:
sure (I think there is a typo, and in a previous revision I'd made substantial modifications here, but I'll put it back for now)

carltongibson reviewed (commented):
Hi @jthorniley — this structure looks much nicer to me. Thanks for making those adjustments. I'll have another closer look. 

We've got time, so I'd like to step through with the tests again (as well). 

Can I ask, did you deploy this/put it under traffic? One assumes it resolves your issue...

jthorniley commented:
> Can I ask, did you deploy this/put it under traffic? One assumes it resolves your issue...

I was just thinking yesterday it would be worth trying that. I should be able to try today and report back.

carltongibson reviewed (approved):
OK, this seems pretty clean to me. 👍 

@jthorniley Did you get that chance to deploy it? 

Thanks for the effort here. 🎁

jthorniley commented:
@carltongibson I did - we've now had a few days running this patch (on top of the 5.0.x branch) with no issues.

nessita commented on django/core/handlers/asgi.py:
```diff
@@ -190,16 +190,28 @@ async def handle(self, scope, receive, send):
+        get_response_task = asyncio.create_task(process_request(request, send))
 
         # Try to catch a disconnect while getting response.
+        disconnect_task = asyncio.create_task(self.listen_for_disconnect(receive))
```
Following the goal of having a minimal (yet correct) patch, would you consider moving this back to the `tasks` list, avoiding the definition of `disconnect_task`?

Also, I'm a bit concerned that changing the order of the tasks in `tasks` does not make any test fail, though that's unrelated to this patch. But if you have any idea on how to test that, we could definitely do a separated PR for it!

nessita commented on tests/asgi/tests.py:
```diff
@@ -376,6 +378,14 @@ async def test_concurrent_async_uses_multiple_thread_pools(self):
+            nonlocal finished_threads
+            finished_threads.append(threading.current_thread())
+
+        request_finished.connect(on_finished)
```
I'm wondering the reason for choosing this approach for recording the `request_finished` thread instead of following the existing code that uses the `SignalHandler`. I know the latter is a bit more... verbose but it would avoid the `nonlocal` call and it will also be more consistent/less surprising when looked in conjunction with the rest of the tests.

jthorniley replied:
Ok sure I can update this to clone the style use in the other tests

nessita reviewed (commented):
Added some comments and many questions :-)

carltongibson commented on django/core/handlers/asgi.py:
```diff
@@ -346,7 +364,6 @@ async def send_response(self, response, send):
                         "more_body": not last,
                     }
                 )
-        await sync_to_async(response.close, thread_sensitive=True)()
```
ASGIHandler is not part of the public API, and this is addressing a bug in a new feature. (I don't see that anyone customising here is covered by any stability guarantees)

nessita replied:
Fair enough! Thanks for your comment.

carltongibson commented on django/core/handlers/asgi.py:
```diff
@@ -190,16 +190,28 @@ async def handle(self, scope, receive, send):
         async def process_request(request, send):
             response = await self.run_get_response(request)
-            await self.send_response(response, send)
+
```
I wouldn't mind if we dropped this blank line.

nessita commented:
@jthorniley Hi! Thank you for your work on this. Would you have time to keep working in this PR? Specifically I would appreciate if you could answer the comments that I added last week so I can progress this work.

jthorniley commented:
@nessita sorry I was a bit slower this week - but I think I've addressed the remaining comments so this is quite a minimal patch now.

jthorniley commented on django/core/handlers/asgi.py:
```diff
@@ -223,6 +229,12 @@ async def process_request(request, send):
                     pass
         body_file.close()
 
+        try:
```
I am now suspecting that this block should go _before_ the call to `body_file.close()`, in case it is still possible for something to try and read from the body file. However there are currently no tests that require this one way or the other (but the old behaviour would have been that the equivalent `request.close()` code would normally have run before `body_file.close()`)

jthorniley replied:
Confirmed: having had this patch rolled out for a while, we were getting a lot of exceptions saying `ValueError: I/O operation on a closed file` that were triggered by web crawler bots (but hadn't been happening under "normal" loads)

I modified the patch so that the new code block goes before the call to `body_file.close()` - and this seems to have fixed the errors.

I don't think there is anything in the tests at the moment that checks for this ordering, but I am confident now that the way it is now is the correct way around to have it.

jthorniley replied:
@nessita I will take a look, it should be possible

jthorniley replied:
This was a little tricky but hopefully `test_cancel_post_request_with_sync_processing` does this. 

If I swap the order of body_file.close() and the request finished block I get a test failure due to the same error being thrown in the test view at the point `request.body` is called.

This maybe seems like an artificial test but I think it is minimal to reproduce the issue I was finding - its quite common for code to be effectively wrapped in sync_to_async due to middleware

nessita replied:
Thank you! I'll review ASAP

nessita commented on tests/asgi/tests.py:
```diff
@@ -194,6 +197,56 @@ async def test_post_body(self):
+        # Record exception raised in the test view
+        exceptions = []
+
+        # This view will run in a new thread because it is wrapped in sync_to_async.
```
Following the [coding style guidelines](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/), "Documentation, comments, and docstrings should be wrapped at 79 characters". Could you please re-wrap these?

nessita commented on tests/asgi/tests.py:
```diff
@@ -194,6 +197,56 @@ async def test_post_body(self):
+        communicator = ApplicationCommunicator(application, scope)
+
+        await communicator.send_input({"type": "http.request", "body": b"Echo!"})
+
```
Could we assert that the response that was returned from the view actually matches `Echo!`? :thinking:

jthorniley replied:
Because we are sending the `http.disconnect` the application never actually returns the response, even though the view code runs concurrently to create it. Perhaps I could put an assertion in the view code, or capture the contents of request.body in there.

jthorniley replied:
I've updated the test so that rather than being an echo, the view just records the content of request_body which can then be asserted at the end of the test

nessita replied:
I like this, thanks!

carltongibson commented on tests/asgi/tests.py:
```diff
@@ -194,6 +197,56 @@ async def test_post_body(self):
         self.assertEqual(response_body["type"], "http.response.body")
         self.assertEqual(response_body["body"], b"Echo!")
 
+    async def test_cancel_post_request_with_sync_processing(self):
```
"expert" is very generous of you 😊

I think this test could do with a docstring. Like I see what it's doing, but just looking at the code, I'm not sure why we're testing this (so I'm going to read the PR discussion). It's likely someone (me) looking in the future could benefit from a steer here. 

The comment on lines 208ff could likely be the most of it, but mentioning that it ensures we've correctly finalised the response before closing the body file (which may be in use) would enable a future reviewer to know which blocks are in play.

jthorniley replied:
I think the headline is that reading request.body should not raise and exception. I've tried to put a brief docstring in to that effect... hopefully this helps

nessita replied:
Given my understanding, this looks good, thanks.

carltongibson commented on tests/asgi/tests.py:
```diff
@@ -194,6 +197,56 @@ async def test_post_body(self):
+        def echo_view(request):
+            try:
+                loop.call_soon_threadsafe(view_started_event.set)
+                time.sleep(0.1)
```
I wonder if we could wait on a `threading.Event` here, rather than sleeping 🤔

jthorniley replied:
I've tried a few things here but I haven't managed to do this with events and keep the test behaviour that I want (the test should fail if I move the body_file.close() in ASGIHandler to before the cleanup code). If using an Event for synchronisation, either the view code gets unblocked too early (and the test doesn't test anything) or I try to put it later and get a deadlock and the test never finishes. While imperfect this might be the only way to get anything tested without making the synchronisation code very complex.

nessita replied:
One thing I would change in this code is having something like an `outcome = []` list to append mappings with the exception and/or body, so when later asserting on its content, the test failure is more informative. Specifically, I would suggest this diff:

```diff
diff --git a/tests/asgi/tests.py b/tests/asgi/tests.py
index 955ec884d4..b562a64c8c 100644
--- a/tests/asgi/tests.py
+++ b/tests/asgi/tests.py
@@ -206,10 +206,8 @@ class ASGITest(SimpleTestCase):
         # Events used to monitor the view processing from the parent test code
         view_started_event = asyncio.Event()
         view_finished_event = asyncio.Event()
-        # Record exception raised in the test view
-        exceptions = []
-        # Request body
-        request_body = []
+        # Record received request body or exceptions raised in the test view
+        outcome = []
 
         # This view will run in a new thread because it is wrapped in
         # sync_to_async. The view consumes the POST body data after a short
@@ -224,10 +222,10 @@ class ASGITest(SimpleTestCase):
                 loop.call_soon_threadsafe(view_started_event.set)
                 time.sleep(0.1)
                 # Do something to read request.body after pause
-                request_body.append(request.body)
+                outcome.append({"request_body": request.body})
                 return HttpResponse("ok")
             except Exception as e:
-                exceptions.append(e)
+                outcome.append({"exception": e})
             finally:
                 loop.call_soon_threadsafe(view_finished_event.set)
 
@@ -257,8 +255,7 @@ class ASGITest(SimpleTestCase):
         with self.assertRaises(asyncio.TimeoutError):
             await communicator.receive_output()
 
-        self.assertEqual(request_body, [b"Body data!"])
-        self.assertEqual(len(exceptions), 0)
+        self.assertEqual(outcome, [{"request_body": b"Body data!"}])
 
     async def test_untouched_request_body_gets_closed(self):
         application = get_asgi_application()
```

That on error produces this failure:

```
AssertionError: Lists differ: [{'exception': ValueError('I/O operation on closed file.')}] != [{'request_body': b'Body data!'}]

First differing element 0:
{'exception': ValueError('I/O operation on closed file.')}
{'request_body': b'Body data!'}

- [{'exception': ValueError('I/O operation on closed file.')}]
+ [{'request_body': b'Body data!'}]
```

Instead of:

```
AssertionError: Lists differ: [] != [b'Body data!']

Second list contains 1 additional elements.
First extra element 0:
b'Body data!'

- []
+ [b'Body data!']
```

carltongibson reviewed (commented):
Hi @nessita — I think the extra test makes sense (and it demonstrates the failure case, moving the `try` block back to the older position, so ... 👍 ) 

I just had a couple of queries.

jthorniley commented on docs/releases/5.0.2.txt:
```diff
@@ -18,3 +18,6 @@ Bugfixes
 
 * Fixed a bug in Django 5.0 that caused a crash of ``Model.full_clean()`` on
   models with a ``GeneratedField`` (:ticket:`35127`).
+* Fixed a regression in Django 5.0 that caused the ``request_finished`` signal to
```
done

jthorniley commented on tests/asgi/tests.py:
```diff
@@ -410,6 +474,10 @@ class TestASGIHandler(ASGIHandler):
+        # Exactly one call to request_finished handler
+        self.assertEqual(len(signal_handler.threads), 1)
+        # It was NOT on the async thread
+        self.assertNotEqual(signal_handler.threads.pop(), threading.current_thread())
```
I've added asserts to that effect

carltongibson commented:
@jthorniley Could I ask you to rebase and push again, to get the CI to go again please? (I wanted to see the failures, but the logs gone already I think.) Thanks

carltongibson commented:
So it looks like the test is hitting a timeout on Python 3.10. 🤔

carltongibson commented:
Hi @jthorniley. 

Tests pass for me with this diff:

```diff
diff --git a/tests/asgi/tests.py b/tests/asgi/tests.py
index b51fa77a7a..955ec884d4 100644
--- a/tests/asgi/tests.py
+++ b/tests/asgi/tests.py
@@ -254,7 +254,7 @@ class ASGITest(SimpleTestCase):
         await communicator.send_input({"type": "http.disconnect"})
         # Wait until view code has finished
         await view_finished_event.wait()
-        with self.assertRaises(TimeoutError):
+        with self.assertRaises(asyncio.TimeoutError):
             await communicator.receive_output()
 
         self.assertEqual(request_body, [b"Body data!"])
```

From [asyncio.TimeoutError docs](https://docs.python.org/3.12/library/asyncio-exceptions.html#asyncio.TimeoutError): 

> Changed in version 3.11: This class was made an alias of `TimeoutError`

nessita commented:
> Hi @jthorniley.
> 
> Tests pass for me with this diff:
> 
> ```diff
> diff --git a/tests/asgi/tests.py b/tests/asgi/tests.py
> index b51fa77a7a..955ec884d4 100644
> --- a/tests/asgi/tests.py
> +++ b/tests/asgi/tests.py
> @@ -254,7 +254,7 @@ class ASGITest(SimpleTestCase):
>          await communicator.send_input({"type": "http.disconnect"})
>          # Wait until view code has finished
>          await view_finished_event.wait()
> -        with self.assertRaises(TimeoutError):
> +        with self.assertRaises(asyncio.TimeoutError):
>              await communicator.receive_output()
>  
>          self.assertEqual(request_body, [b"Body data!"])
> ```
> 
> From [asyncio.TimeoutError docs](https://docs.python.org/3.12/library/asyncio-exceptions.html#asyncio.TimeoutError):
> 
> > Changed in version 3.11: This class was made an alias of `TimeoutError`

This is amazing investigation, thank you @carltongibson. Did it take you too long to get to this result? With the fix spelled out, it seems simple, but I'm sure this was a rabbit hole until you got there!

nessita commented:
@jthorniley Let me know if you can work on the pending items (which are only a few), otherwise no worries I'll push this to the finish line tomorrow or the day after. Thank you both!

jthorniley commented:
We also have a conflict in the release notes file so I'll try to rebase and fix that

nessita commented:
> I think I can see outstanding the python 3.10 diff which should be trivial, and also rewriting the try-except as try-except-else - I should be able to do both these this afternoon.

Great! Thank you!

> I'm equally happy for you to take over this PR if you see any other bits that need tidying up.

We try to encourage and prioritize work from the community, so I only offered to avoid putting pressure on you -- everything you can do yourself helps and is very welcomed!

nessita reviewed (commented):
@jthorniley Branch looks great, thank you. I added a minor comment (that we don't necessarily need to apply).

What I would need is some answers to my points (and sub points) 1 and 2 from my previous review. Do you think you could help me understand those?

nessita commented:
> I've reviewed the code and added some comments. Thanks for the work so far! I think this branch looks good and once merged I'll backport to stable/5.0.x branch.
> 
> Having said that, I have some questions to try to understand the bigger picture in this handler:
> 
>     1. What happens if `send_response` raises an exception in the `if request is None` branch?
>        1.1. In this same `if` branch, would `body_file.close()` also potentially raise `ValueError: I/O operation on a closed file` since it's called before sending the response?
>        1.2. In this same `if` branch, I think there is still an unanswered question about `response.close` not being called (since was called before this change).
> 
>     2. Is there a scenario where `run_get_response` could also raise `CancelledError` (since it awaits on `self.get_response_async`)?

@jthorniley As I just mentioned in my review submission, if you could share some thoughts on the above, it would help me finalize the review and merge this PR. Thank you!

carltongibson commented:
Hi @nessita. 

For query 1, that code is essentially unchanged since it was introduced (a415ce7) but see f1e0fc6, which made sure we close the `body_file` on that block. If there's a problem there I would have expected it to come up already, but it's out of scope for the release blocker here no? 

For query 2, `CancelledError` is _raised into_ the suspended coroutine at the point where the cancelation is received. If that cancelation happens during `run_get_response` we won't get the response (the coroutine simply never completes), and that's where we need the final `else` block to dispatch `request_finished` at the end. (Often you don't handle the `CancelledError`; normally you'd re-raise it, but we wanted to return the `response` if we had one, in order to go via `response.close`, rather than using the bare `request_finished`.)

jthorniley commented:
Sorry @nessita I think I lost these questions among the various other updates I've been making

> * What happens if `send_response` raises an exception in the `if request is None` branch?

I think following @carltongibson comment - this behaviour should stay the same here: before and after this PR, the response.close() cleanup code would not be called if an exception is raised here. 

`if request is None` here means that the `create_request` function caught one of two exceptions and is saying that the response should be 400 or 413 immediately before any view processing. 

>   1.1. In this same `if` branch, would `body_file.close()` also potentially raise `ValueError: I/O operation on a closed file` since it's called before sending the response?

I think that error occurs if a body_file.read() happens after the body_file.close(), since we "know" that the response here is a fixed 400 or 413 returned by `create_request`, without any view processing that would want to read from the POST data, this won't happen.

>   1.2. In this same `if` branch, I think there is still an unanswered question about `response.close` not being called (since was called before this change).

If `send_response` doesn't create an exception yes this is a change of behaviour, I think the reasoning was that it wasn't clear that the `response.close()` in this instance was _intentional_ in the previous code or just a side-effect of it being at the end of `send_response` - there was no test case covering it and I am yet to find any downside of not calling it.

> * Is there a scenario where `run_get_response` could also raise `CancelledError` (since it awaits on `self.get_response_async`)?

Yes - I think that is possible and this would be passed to the calling code when tasks[1].result() is called on line 232, and handled on line 233/4. In this case, the cancellation has occurred before the `response` object is available which is why I directly call the request_finished signal instead.

nessita commented:
> Hi @nessita.
> 
> For query 1, that code is essentially unchanged since it was introduced ([a415ce7](https://github.com/django/django/commit/a415ce70bef6d91036b00dd2c8544aed7aeeaaed)) but see f1e0fc6, which made sure we close the `body_file` on that block. If there's a problem there I would have expected it to come up already, but it's out of scope for the release blocker here no?

Indeed you are correct that any potential bugs/fixes in this branch are out of scope. Though this PR does introduces a logic change in this exact code block: currently, `send_response` would call `response.close` and in this PR it stops doing so (see related response from James about this, he acknowledges the logic change but suspects that it's OK to not call `response.close` for the `error_response` being returned).

> For query 2, `CancelledError` is _raised into_ the suspended coroutine at the point where the cancelation is received. If that cancelation happens during `run_get_response` we won't get the response (the coroutine simply never completes), and that's where we need the final `else` block to dispatch `request_finished` at the end. (Often you don't handle the `CancelledError`; normally you'd re-raise it, but we wanted to return the `response` if we had one, in order to go via `response.close`, rather than using the bare `request_finished`.)

This makes sense, thank you for the thorough explanation!

carltongibson commented:
> Though this PR does introduces a logic change in this exact code block: currently, send_response would call response.close and in this PR it stops doing so

Yes! Good point. I missed that in the discussion. 

Likely we **should** call `error_response.close()` even in this case 🤔

nessita commented:
> > Though this PR does introduces a logic change in this exact code block: currently, send_response would call response.close and in this PR it stops doing so
> 
> Yes! Good point. I missed that in the discussion.
> 
> Likely we **should** call `error_response.close()` even in this case 🤔

Great! I tried to put together a test for this, following @jthorniley explanation that this code block (`if request is None: ...`) follows a failure to create the request in `self.create_request(...)`, which is likely due to the view returning a 40x response. So I wanted to create a `test_post_error` (following the existing `test_post_body`), but I'm still entangled in understanding how to use `ApplicationCommunicator` to do this.

I'm reading [these docs](https://channels.readthedocs.io/en/stable/topics/testing.html) but I'm yet not sure how to pass/define a specific view to return a non 200 response. I also see that this PR, in its new `test_cancel_post_request_with_sync_processing` test, creates a `TestASGIHandler`  with a custom `TestASGIRequest` which defines a custom `urlconf` pointing to a custom `post_view`. Is this the "right way" of doing so? I was hoping for a more straightforward way to use a custom failing view.

Any pointer in this direction will be appreciated!

nessita commented:
@carltongibson Thank you! This is what I came up with, if you +1 I'll incorporate to the PR and merge :woman_dancing: 

```diff
diff --git a/django/core/handlers/asgi.py b/django/core/handlers/asgi.py
index 7e7d5a4aae..bf3ef4a031 100644
--- a/django/core/handlers/asgi.py
+++ b/django/core/handlers/asgi.py
@@ -186,6 +186,7 @@ class ASGIHandler(base.BaseHandler):
         if request is None:
             body_file.close()
             await self.send_response(error_response, send)
+            await sync_to_async(error_response.close)()
             return
 
         async def process_request(request, send):
diff --git a/tests/asgi/tests.py b/tests/asgi/tests.py
index b562a64c8c..baae30b11a 100644
--- a/tests/asgi/tests.py
+++ b/tests/asgi/tests.py
@@ -9,6 +9,7 @@ from asgiref.testing import ApplicationCommunicator
 
 from django.contrib.staticfiles.handlers import ASGIStaticFilesHandler
 from django.core.asgi import get_asgi_application
+from django.core.exceptions import RequestDataTooBig
 from django.core.handlers.asgi import ASGIHandler, ASGIRequest
 from django.core.signals import request_finished, request_started
 from django.db import close_old_connections
@@ -197,6 +198,45 @@ class ASGITest(SimpleTestCase):
         self.assertEqual(response_body["type"], "http.response.body")
         self.assertEqual(response_body["body"], b"Echo!")
 
+    async def test_create_request_error(self):
+
+        class SignalHandler:
+            """Track threads handler is dispatched on."""
+
+            threads = []
+
+            def __call__(self, **kwargs):
+                self.threads.append(threading.current_thread())
+
+        signal_handler = SignalHandler()
+        request_finished.connect(signal_handler)
+
+        # Request class that always fails creation with RequestDataTooBig
+        class TestASGIRequest(ASGIRequest):
+
+            def __init__(self, scope, body_file):
+                super().__init__(scope, body_file)
+                raise RequestDataTooBig()
+
+        # Handler to use request class.
+        class TestASGIHandler(ASGIHandler):
+            request_class = TestASGIRequest
+
+        application = TestASGIHandler()
+        scope = self.async_request_factory._base_scope(
+            method="POST",
+            path="/post/",
+        )
+        communicator = ApplicationCommunicator(application, scope)
+
+        # Initiate request
+        await communicator.send_input({"type": "http.request"})
+        # Give response.close() time to finish.
+        await communicator.wait()
+
+        self.assertEqual(len(signal_handler.threads), 1)
+        self.assertNotEqual(signal_handler.threads[0], threading.current_thread())
+
     async def test_cancel_post_request_with_sync_processing(self):
         """
         The request.body object should be available and readable in view
```

carltongibson commented:
@nessita I didn't run it, but that looks correct to me 👍 

Good job keeping track of the outstanding points here! 🎁

nessita commented:
> @nessita I didn't run it, but that looks correct to me 👍

Amazing, I did run it and I double checked it fails if we remove the `await sync_to_async(error_response.close)()` call. Failure in that case is:

```
$ ./runtests.py asgi.tests.ASGITest.test_create_request_error
Testing against Django installed in '/home/nessita/fellowship/django/django' with up to 4 processes
Found 1 test(s).
System check identified no issues (0 silenced).
F
======================================================================
FAIL: test_create_request_error (asgi.tests.ASGITest.test_create_request_error)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/nessita/.virtualenvs/djangodev/lib/python3.11/site-packages/asgiref/sync.py", line 277, in __call__
    return call_result.result()
           ^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/concurrent/futures/_base.py", line 449, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/home/nessita/.virtualenvs/djangodev/lib/python3.11/site-packages/asgiref/sync.py", line 353, in main_wrap
    result = await self.awaitable(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/nessita/fellowship/django/tests/asgi/tests.py", line 237, in test_create_request_error
    self.assertEqual(len(signal_handler.threads), 1)
AssertionError: 0 != 1

----------------------------------------------------------------------
Ran 1 test in 0.004s

FAILED (failures=1)
```

> Good job keeping track of the outstanding points here! 🎁 (This being comment 104)

:heart: 

Will push and merge soon!

carltongibson commented:
Looks like we're there. Thanks for the report and work on getting this 100% right @jthorniley 🏅

nessita commented:
@jthorniley Amazing collaborative work, thank you so much.

I added the removed `error_response.close` call with a test, and made final tweaks. I pushed two commits, the first one moving the `SignalHandler` helper class to be at module level so it can be reused by the multiple tests.

Let me know what you think!

jthorniley commented:
@nessita @carltongibson thanks for the updates - just catching up on the conversation.

So I think the summary is we have added back a call to `request.close()` specifically in that early error case (`if request is None`), because that minimises the difference in behaviour before/after this PR, and we've added a test case to track that behaviour as it was evidently not covered before. 

This all makes sense to me so I'm happy with that being added.
